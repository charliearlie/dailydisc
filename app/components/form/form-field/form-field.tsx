import { type LucideIcon } from "lucide-react";
import { type ReactElement, forwardRef, useId } from "react";

import { Input } from "~/components/common/ui/input";
import { Label } from "~/components/common/ui/label";
import { InputErrors } from "~/components/form/input-errors";

import { cn } from "~/util/utils";

export type FormFieldProps = {
  errors?: Array<string> | null;
  helperText?: string;
  Icon?: LucideIcon;
  label: string;
};

type Props = FormFieldProps & React.HTMLProps<HTMLInputElement>;

export const FormField = forwardRef<HTMLInputElement, Props>(
  (
    { errors, helperText, Icon, label, name, type, ...props },
    ref
  ): ReactElement => {
    const id = useId();
    const inputId = `${id}-${name}`;
    const inputErrorsId = `${id}-${name}-errors`;

    const hasErrors = !!errors?.length;
    return (
      <div className="flex w-full flex-col gap-1.5">
        <Label className="font-bold" htmlFor={inputId}>
          {label}
        </Label>
        <span className="flex items-center">
          <Input
            name={name}
            className={cn(
              hasErrors && "ring-2 ring-destructive ring-offset-1",
              Icon && "pl-5",
              type === "file" &&
                "file:-mx-3 file:-my-[0.32rem] file:cursor-pointer file:overflow-hidden file:rounded-none file:border-0 file:border-solid file:border-inherit file:bg-background file:px-3 file:py-[0.32rem] file:text-foreground file:transition file:duration-150 file:ease-in-out file:[margin-inline-end:0.75rem] file:[border-inline-end-width:3px] hover:file:opacity-70"
            )}
            id={inputId}
            type={type}
            {...props}
            ref={ref}
            aria-invalid={hasErrors || undefined}
            aria-describedby={hasErrors ? inputErrorsId : undefined}
          />
          {Icon && <Icon className="absolute" strokeWidth={3} size={16} />}
        </span>
        {helperText && !hasErrors && (
          <span className="text-sm font-light">{helperText}</span>
        )}
        <div className="flex min-h-[18px] items-start">
          <InputErrors id={inputErrorsId} errors={errors} />
        </div>
      </div>
    );
  }
);

FormField.displayName = "FormField";
