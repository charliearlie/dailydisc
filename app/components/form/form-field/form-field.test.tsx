import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

import { FormField } from "./form-field";

describe("FormField", () => {
  it("renders the label and input correctly", () => {
    const label = "Username";
    const inputName = "username";

    const { container, getByLabelText } = render(
      <FormField label={label} name={inputName} />
    );

    const input = getByLabelText(label);

    expect(input).toHaveAttribute("name", inputName);
    expect(container).toMatchSnapshot();
  });

  it("renders the errors correctly", () => {
    const label = "Username";
    const inputName = "username";
    const errors = [
      "Username is required",
      "Username must be at least 6 characters long",
    ];
    const { getByText } = render(
      <FormField label={label} name={inputName} errors={errors} />
    );
    errors.forEach((error) => {
      expect(getByText(error)).toBeInTheDocument();
    });
  });

  it("should put the input into an error state when there are errors", () => {
    const label = "Username";
    const inputName = "username";
    const errors = [
      "Username is required",
      "Username must be at least 6 characters long",
    ];
    const { getByLabelText } = render(
      <FormField label={label} name={inputName} errors={errors} />
    );

    const inputElement = getByLabelText(label);

    expect(inputElement).toHaveAttribute("aria-invalid", "true");
  });
});
