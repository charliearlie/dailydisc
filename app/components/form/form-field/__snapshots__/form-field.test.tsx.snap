// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders the label and input correctly 1`] = `
<div>
  <div
    class="flex w-full flex-col gap-1.5"
  >
    <label
      class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 font-bold"
      for=":r0:-username"
    >
      Username
    </label>
    <span
      class="flex items-center"
    >
      <input
        class="flex h-10 w-full rounded-md border border-input bg-input px-3 py-2 ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-username"
        name="username"
      />
    </span>
    <div
      class="flex min-h-[18px] items-start"
    />
  </div>
</div>
`;
